src/soc/ibm/power9/powerbus.c

```diff
@@ -0,0 +1,295 @@
+/* SPDX-License-Identifier: GPL-2.0-only */
+
+#include <cpu/power/powerbus.h>
+
+#include <cpu/power/mvpd.h>
+#include <cpu/power/scom.h>
+#include <console/console.h>
+#include <stdint.h>
+
+#define EPSILON_MAX_VALUE 0xFFFFFFFF
+
+#define VOLTAGE_DATA_VERSION 3
+#define VOLTAGE_BUCKET_COUNT 6
+
+#define MBOX_SCRATCH_REG1 0x00050038
+#define MBOX_SCRATCH_REG6_GROUP_PUMP_MODE (1 << 23)
+
+/*
+ * EPS table type.
+ * From src/import/chips/p9/procedures/xml/attribute_info/nest_attributes.xml
+ */
+#define EPS_TYPE_LE	0x01
+#define EPS_TYPE_HE	0x02
+#define EPS_TYPE_HE_F8	0x03
+
+/* Single bucket within #V keyword of version 3 */
+struct voltage_data
+{
+	uint16_t freq;
+	uint16_t vdd_voltage;
+	uint16_t idd_current;
+	uint16_t vcs_voltage;
+	uint16_t ics_current;
+} __attribute__((__packed__));
+
+/* Single bucket within #V keyword of version 3 */
+struct voltage_bucket_data
+{
+	uint8_t id;
+
+	struct voltage_data nominal;
+	struct voltage_data powersave;
+	struct voltage_data turbo;
+	struct voltage_data ultra_turbo;
+	struct voltage_data powerbus;
+
+	uint16_t sort_power_normal;
+	uint16_t sort_power_turbo;
+
+	uint8_t reserved[6];
+} __attribute__((__packed__));
+
+struct voltage_kwd
+{
+	uint8_t version;
+	uint8_t pnp[3];
+	struct voltage_bucket_data buckets[VOLTAGE_BUCKET_COUNT];
+} __attribute__((__packed__));
```

Comment:
Or maybe even rename `powerbus_cfg` and save whole bucket

Reply:
That would be nice. In 15.1 I need all `voltage_data` for cores, bucket ID and PB frequency. I may also need those two `sort_power_*` fields because they are named similar to the attribute I'll have to find. I don't think they differ between quads, but can't tell for sure.
The logic in 15.1 is strange - it iterates over buckets and compares PB freq in bucket with attribute, which was created based on the same MVPD data, so it will match first non-zero PB freq anyway. Hostboot keeps surprising me in a bad way.